staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/validator_test.go

```diff
@@ -572,6 +581,244 @@ func TestValidate(t *testing.T) {
+				{
+					EvalResult: celtypes.False,
+					ExpressionAccessor: &ValidationCondition{
+						Reason:     &forbiddenReason,
+						Expression: "this.expression == unit.test",
+						Message:    "test1",
+					},
+				},
+			},
+			messageEvaluations: []cel.EvaluationResult{
+				{
+					Error: &apiservercel.Error{Type: apiservercel.ErrorTypeInvalid},
+				},
+			},
+			policyDecision: []PolicyDecision{
+				{
+					Action:  ActionDeny,
+					Message: "test1", // original message used
+					Reason:  forbiddenReason,
+				},
+			},
+		},
+		{
+			name: "messageExpression eval to empty",
+			evaluations: []cel.EvaluationResult{
+				{
+					EvalResult: celtypes.False,
+					ExpressionAccessor: &ValidationCondition{
+						Reason:     &forbiddenReason,
+						Expression: "this.expression == unit.test",
+						Message:    "test1",
+					},
+				},
+			},
+			messageEvaluations: []cel.EvaluationResult{
+				{
+					EvalResult: celtypes.String(" "),
+				},
+			},
+			policyDecision: []PolicyDecision{
+				{
+					Action:  ActionDeny,
+					Message: "test1",
+					Reason:  forbiddenReason,
+				},
+			},
+		},
+		{
+			name: "messageExpression eval to multi-line",
+			evaluations: []cel.EvaluationResult{
+				{
+					EvalResult: celtypes.False,
+					ExpressionAccessor: &ValidationCondition{
+						Reason:     &forbiddenReason,
+						Expression: "this.expression == unit.test",
+						Message:    "test1",
+					},
+				},
+			},
+			messageEvaluations: []cel.EvaluationResult{
+				{
+					EvalResult: celtypes.String("hello\nthere"),
+				},
+			},
+			policyDecision: []PolicyDecision{
+				{
+					Action:  ActionDeny,
+					Message: "test1",
+					Reason:  forbiddenReason,
+				},
+			},
+		},
+		{
+			name: "messageExpression eval result too long",
+			evaluations: []cel.EvaluationResult{
+				{
+					EvalResult: celtypes.False,
+					ExpressionAccessor: &ValidationCondition{
+						Reason:     &forbiddenReason,
+						Expression: "this.expression == unit.test",
+						Message:    "test1",
+					},
+				},
+			},
+			messageEvaluations: []cel.EvaluationResult{
+				{
+					EvalResult: celtypes.String(strings.Repeat("x", 5*1024+1)),
+				},
+			},
+			policyDecision: []PolicyDecision{
+				{
+					Action:  ActionDeny,
+					Message: "test1",
+					Reason:  forbiddenReason,
+				},
+			},
+		},
+		{
+			name: "messageExpression eval to null",
+			evaluations: []cel.EvaluationResult{
+				{
+					EvalResult: celtypes.False,
+					ExpressionAccessor: &ValidationCondition{
+						Reason:     &forbiddenReason,
+						Expression: "this.expression == unit.test",
+						Message:    "test1",
+					},
+				},
+			},
+			messageEvaluations: []cel.EvaluationResult{
+				{
+					EvalResult: celtypes.NullValue,
+				},
+			},
+			policyDecision: []PolicyDecision{
+				{
+					Action:  ActionDeny,
+					Message: "test1",
+					Reason:  forbiddenReason,
+				},
+			},
+		},
+		{
+			name: "messageExpression out of budget after successful eval of expression",
```

Comment:
Great to see the test for budget shared between expression and messageExpression

Reply:
:)